src/Geometry/RefinedDiscreteModels.jl

```diff
@@ -0,0 +1,298 @@
+  for k = 1:NE
+    i = edge[k, 1]
+    j = edge[k, 2]
+    d2p[i, j] = k
+    d2p[j, i] = k
+  end
+  d2p
+end
+
+function test_against_top(face::Matrix{<:Integer}, top::GridTopology, d::Integer)
+  face_vec = sort.([face[i, :] for i = 1:size(face, 1)])
+  face_top = sort.(get_faces(top, d, 0))
+  issetequal_bitvec = issetequal(face_vec, face_top)
+  @assert all(issetequal_bitvec)
+end
+
+function get_midpoint(ngon::AbstractArray{<:VectorValue})
+  sum(ngon) / length(ngon)
+end
+
+function Base.atan(v::VectorValue{2,T}) where {T<:AbstractFloat}
+  atan(v[2], v[1])
+end
+
+function Base.:^(v::VectorValue{N,T}, r::Integer) where {T, N}
+  VectorValue([v[i]^r for i = 1:N]...)
+end
+
+function sort_ccw(cell_coords::AbstractArray{<:VectorValue})
+  midpoint = get_midpoint(cell_coords)
+  offset_coords = cell_coords .- midpoint
+  sortperm(offset_coords, by = atan)
+end
+
+function sort_cell_node_ids_ccw(
+    cell_node_ids::Table{<:Integer},
+    node_coords::AbstractArray{<:VectorValue},
+  )
+  cell_node_ids_ccw = vcat(cell_node_ids'...)
+  #cell_node_ids_ccw = cell_node_ids
+  #@show cell_node_ids_ccw
+  for (i, cell) in enumerate(cell_node_ids)
+    cell_coords = node_coords[cell]
+    perm = sort_ccw(cell_coords)
+    cell_node_ids_ccw[i, :] = cell[perm]
+  end
+  cell_node_ids_ccw
+end
+
+"""
+node_coords == node, cell_node_ids == elem in Long Chen's notation
+"""
+function newest_vertex_bisection(
+  top::GridTopology,
+  node_coords::AbstractArray{<:VectorValue},
+  cell_node_ids::Matrix{Ti},
+  η_arr::AbstractArray{<:AbstractFloat},
+  θ::AbstractFloat,
+  sort_flag::Bool,
+) where {Ti <: Integer}
+  # Number of nodes
+  N::Ti = size(node_coords, 1)
+  elem = cell_node_ids
+  # Number of cells (triangles in 2D)
+  NT::Ti = size(elem, 1)
+  @assert length(η_arr) == NT
+  if sort_flag
+    sort_longest_edge!(elem, node_coords, NT)
+  end
+  # Make sure elem is consistent with GridTopology
+  test_against_top(elem, top, 2)
+  edge = build_edges(elem)
+  NE::Ti = size(edge, 1)
+  dualedge = build_directed_dualedge(elem, N, NT)
+  d2p = dual_to_primal(edge, NE, N)
+  # Make sure edge is consistent with GridTopology
+  test_against_top(edge, top, 1)
+  node_coords, marker =
+    setup_markers_and_nodes!(node_coords, elem, d2p, dualedge, NE, η_arr, θ)
+  cell_node_ids = bisect(d2p, elem, marker, NT)
+  node_coords, cell_node_ids
+end
+
+# step 1
+function newest_vertex_bisection(
+  grid::Grid,
+  top::GridTopology,
+  η_arr::AbstractArray{<:AbstractFloat},
+  θ::AbstractFloat,
+  sort_flag::Bool,
+)
+  node_coords = get_node_coordinates(grid)
+  cell_node_ids = get_cell_node_ids(grid)
+  if sort_flag
+    cell_node_ids_ccw = sort_cell_node_ids_ccw(cell_node_ids, node_coords)
+  else
+    cell_node_ids_ccw = vcat(cell_node_ids'...)
+  end
+  node_coords_ref, cell_node_ids_ref =
+    newest_vertex_bisection(top, node_coords, cell_node_ids_ccw, η_arr, θ, sort_flag)
+  # TODO: Should not convert to matrix and back to Table
```

Comment:
`cell_node_ids_ref` can be built as a Table from the beginning.

Reply:
>We also need to mutate Tables. To this end, we access the fields .data and .ptrs from the Table object directly . They are such that a[i][j] == a.data[ a.ptrs[i]-1+j ] for a::Table. This is a pretty standard way of storing a vector of vectors with 2 consecutive chunks of memory. It it also related with sparse matrix CSR/CSC storage. You are provably already familiar with this.

Yes, I am familiar with this storage format. This works great! I am able to modify the rows of a `Table` now. However, I am still seeing a large memory footprint when I add a new row to a `Table` using `append_tables_globally` via something like

```julia
new_row = #some vector
elem = append_tables_globally(elem, Table([new_row])) 
```
is there a better way to achieve this? I saw some `copy`s in the definition of `append_tables_globally` for this operation. But maybe it's unavoidable since we don't know the memory size a priori. In this case though, I do know beforehand that I am just appending a new row though..

Reply:
> But maybe it's unavoidable since we don't know the memory size a priori.

If you don't know the memory size a priori, then perhaps the best option is to work with a vector of vectors and use push! And finally convert this to a Table just before building the UnstructuredGrid. Function append_tables_globally is clearly not designed to be called at each cell.